core/src/main/java/io/grpc/util/SubchannelListLoadBalancerCommons.java

```diff
@@ -0,0 +1,348 @@
+  }
+
+  /**
+   * Common new addresses list update handling.
+   */
+  public boolean acceptResolvedAddresses(ResolvedAddresses resolvedAddresses) {
+    if (resolvedAddresses.getAddresses().isEmpty()) {
+      handleNameResolutionError(Status.UNAVAILABLE.withDescription(
+          "NameResolver returned no usable address. addrs=" + resolvedAddresses.getAddresses()
+              + ", attrs=" + resolvedAddresses.getAttributes()));
+      return false;
+    }
+
+    List<EquivalentAddressGroup> servers = resolvedAddresses.getAddresses();
+    Set<EquivalentAddressGroup> currentAddrs = subchannels.keySet();
+    Map<EquivalentAddressGroup, EquivalentAddressGroup> latestAddrs = stripAttrs(servers);
+    Set<EquivalentAddressGroup> removedAddrs = setsDifference(currentAddrs, latestAddrs.keySet());
+
+    for (Map.Entry<EquivalentAddressGroup, EquivalentAddressGroup> latestEntry :
+        latestAddrs.entrySet()) {
+      EquivalentAddressGroup strippedAddressGroup = latestEntry.getKey();
+      EquivalentAddressGroup originalAddressGroup = latestEntry.getValue();
+      Subchannel existingSubchannel = subchannels.get(strippedAddressGroup);
+      if (existingSubchannel != null) {
+        // EAG's Attributes may have changed.
+        existingSubchannel.updateAddresses(Collections.singletonList(originalAddressGroup));
+        continue;
+      }
+      // Create new subchannels for new addresses.
+
+      // NB(lukaszx0): we don't merge `attributes` with `subchannelAttr` because subchannel
+      // doesn't need them. They're describing the resolved server list but we're not taking
+      // any action based on this information.
+      Attributes.Builder subchannelAttrs = Attributes.newBuilder()
+          // NB(lukaszx0): because attributes are immutable we can't set new value for the key
+          // after creation but since we can mutate the values we leverage that and set
+          // AtomicReference which will allow mutating state info for given channel.
+          .set(STATE_INFO,
+              new Ref<>(ConnectivityStateInfo.forNonError(IDLE)));
+
+      final Subchannel subchannel = checkNotNull(
+              createSubchannelTask.apply(CreateSubchannelArgs.newBuilder()
+              .setAddresses(originalAddressGroup)
+              .setAttributes(subchannelAttrs.build())
+              .build()),
+          "subchannel");
+      subchannel.start(new SubchannelStateListener() {
+          @Override
+          public void onSubchannelState(ConnectivityStateInfo state) {
+            processSubchannelState(subchannel, state);
+          }
+        });
+      subchannels.put(strippedAddressGroup, subchannel);
+      subchannel.requestConnection();
+    }
+
+    ArrayList<Subchannel> removedSubchannels = new ArrayList<>();
+    for (EquivalentAddressGroup addressGroup : removedAddrs) {
+      removedSubchannels.add(subchannels.remove(addressGroup));
+    }
+
+    afterUpdateTask.run();
```

Comment:
What will this be used for?

Reply:
For OOB listeners.  I wanted to add/remove listeners on the subchannels once new config about `enableOobLoadReport` is received and I want to do this before any picker activity, e.g. updateSubchannelStatus() so the new weight would come before than a new picker might be created.  But there won't be new weight if subchannel is not ready yet.
Alternatively I can apply the config after the entire `acceptResolvedAddress` then I don't need to add this task here,  but this might cause small time window when both OOB and per-request are effective, or we miss the initial reports.

Reply:
I see. I think in that case we could just do it afterward and "let things happen." If we have both oob and in-band, then... nothing's really different. No matter what we'll be taking the most recent report and throwing away the earlier ones. Starting OOB also takes a long time relative to this code, as it actually has to do an RPC.

But I do also see how we could have situations where this position is important, in the same way it is important for updating the picker. But I also think we need to be thinking about the interactions with config update and the picker update at the same time.